File: construction_heuristic.cpp
#include <vector>
#include <numeric>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <omp.h>
#include <nlohmann/json.hpp>
#include <CLI/CLI.hpp>

using json = nlohmann::json;

struct HeuristicResult {
    std::vector<std::vector<std::vector<double>>> x;
    std::vector<std::vector<double>> y;
    std::vector<std::vector<std::vector<double>>> w;
    std::vector<std::vector<double>> q;
    std::vector<std::vector<double>> z;
};

HeuristicResult construction_heuristic(int nF, int nC, int nT, 
                                       const std::vector<std::vector<double>>& d, 
                                       const std::vector<double>& q_in) {
    
    // Initialize variables
    std::vector<std::vector<std::vector<double>>> x(nF, std::vector<std::vector<double>>(nC, std::vector<double>(nT, 0.0)));
    std::vector<std::vector<double>> y(nF, std::vector<double>(nT, 1.0));
    std::vector<std::vector<std::vector<double>>> w(nF, std::vector<std::vector<double>>(nF, std::vector<double>(nT, 0.0)));
    std::vector<std::vector<double>> q(nF, std::vector<double>(nT));
    std::vector<std::vector<double>> z(nF, std::vector<double>(nT, 0.0));
    
    // Initialize q by tiling q_in
    for (int f = 0; f < nF; ++f) {
        for (int t = 0; t < nT; ++t) {
            q[f][t] = q_in[f];
        }
    }
    
    // Check for feasibility
    double q_in_sum = std::accumulate(q_in.begin(), q_in.end(), 0.0);
    for (int t = 0; t < nT; ++t) {
        double demand_sum = 0.0;
        for (int c = 0; c < nC; ++c) {
            demand_sum += d[c][t];
        }
        if (q_in_sum < demand_sum) {
            std::cout << "Not enough resources for demand at time " << t 
                      << ". Problem is infeasible." << std::endl;
            std::exit(1);
        }
    }
    
    // Main assignment loop - PARALLELIZED
    #pragma omp parallel for schedule(dynamic)
    for (int t = 0; t < nT; ++t) {
        for (int c = 0; c < nC; ++c) {
            double already_assigned = 0.0;
            
            int f = 0;
            double sum_x_f = 0.0;
            for (int k = 0; k < nC; ++k) {
                sum_x_f += x[f][k][t];
            }
            double residual_capacity = q_in[f] - sum_x_f;
            
            while (already_assigned < d[c][t]) {
                if (residual_capacity > d[c][t] - already_assigned) {
                    x[f][c][t] += d[c][t] - already_assigned;
                    already_assigned = d[c][t];
                } else {
                    if (residual_capacity > 0) {
                        x[f][c][t] += residual_capacity;
                        already_assigned += residual_capacity;
                    }
                    f++;
                    if (f < nF) {
                        sum_x_f = 0.0;
                        for (int k = 0; k < nC; ++k) {
                            sum_x_f += x[f][k][t];
                        }
                        residual_capacity = q_in[f] - sum_x_f;
                    }
                }
            }
        }
    }
    
    // Normalization loop - PARALLELIZED
    // #pragma omp parallel for schedule(dynamic)
    for (int t = 0; t < nT; ++t) {
        for (int j = 0; j < nC; ++j) {
            double sum_x_j = 0.0;
            for (int i = 0; i < nF; ++i) {
                sum_x_j += x[i][j][t];
            }
            
            if (sum_x_j != 0.0) {
                for (int i = 0; i < nF; ++i) {
                    x[i][j][t] = x[i][j][t] / sum_x_j;
                }
            }
        }
    }
    
    return {x, y, w, q, z};
}

// Read input from JSON file
bool read_input(const std::string& filename, 
                int& nF, int& nC, int& nT,
                std::vector<std::vector<double>>& d,
                std::vector<double>& q_in) {
    try {
        std::ifstream input_file(filename);
        if (!input_file.is_open()) {
            std::cerr << "Error: Cannot open input file " << filename << std::endl;
            return false;
        }
        
        json j;
        input_file >> j;
        
        nF = j["nF"];
        nC = j["nC"];
        nT = j["nT"];
        
        // Read d (nC x nT matrix)
        d = j["d"].get<std::vector<std::vector<double>>>();
        
        // Read q_in (nF vector)
        q_in = j["q_in"].get<std::vector<double>>();
        
        //std::cout << "Input loaded successfully: nF=" << nF 
        //          << ", nC=" << nC << ", nT=" << nT << std::endl;
        
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Error reading input file: " << e.what() << std::endl;
        return false;
    }
}

// Write output to JSON file
bool write_output(const std::string& filename, const HeuristicResult& result) {
    try {
        json j;
        
        j["x"] = result.x;
        j["y"] = result.y;
        j["w"] = result.w;
        j["q"] = result.q;
        j["z"] = result.z;
        
        std::ofstream output_file(filename);
        if (!output_file.is_open()) {
            std::cerr << "Error: Cannot open output file " << filename << std::endl;
            return false;
        }
        
        // Write with indentation for readability
        output_file << j.dump(2);
        
        // std::cout << "Output written successfully to " << filename << std::endl;
        
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Error writing output file: " << e.what() << std::endl;
        return false;
    }
}

int main(int argc, char** argv) {
    // Setup CLI
    CLI::App app{"Construction Heuristic Solver"};
    
    // Command-line arguments
    std::string input_file;
    std::string output_file;
    int num_threads = 0;  // 0 means use default/all available
    
    app.add_option("-i,--input", input_file, "Input JSON file path")
        ->required()
        ->check(CLI::ExistingFile);
    
    app.add_option("-o,--output", output_file, "Output JSON file path")
        ->required();
    
    app.add_option("-t,--threads", num_threads, "Number of OpenMP threads (default: all available)")
        ->check(CLI::PositiveNumber);
    
    // Parse command-line arguments
    CLI11_PARSE(app, argc, argv);
    
    // Variables
    int nF, nC, nT;
    std::vector<std::vector<double>> d;
    std::vector<double> q_in;
    
    // Read input
    if (!read_input(input_file, nF, nC, nT, d, q_in)) {
        return 1;
    }
    
    // Set number of threads
    if (num_threads > 0) {
        omp_set_num_threads(num_threads);
        //std::cout << "Using " << num_threads << " threads" << std::endl;
    } /* else {
        std::cout << "Using default number of threads" << std::endl;
    }*/
    
    // Run heuristic
    // std::cout << "Running construction heuristic..." << std::endl;
    auto result = construction_heuristic(nF, nC, nT, d, q_in);
    
    // Write output
    if (!write_output(output_file, result)) {
        return 1;
    }
    
    // std::cout << "Done!" << std::endl;
    return 0;
}

File: construction_heuristic.py
import subprocess
import json
import os
import sys
from pathlib import Path
import time
import numpy as np
import logging

# from corberan_data import CorberanData
import toy_instances as toy

EXECUTABLE_PATH = Path(__file__).parent / "construction_heuristic"


def construction_heuristic(nF, nC, nT, d, q_in):

    # Variables
    x = np.zeros((nF, nC, nT))
    y = np.ones((nF, nT))
    w = np.zeros((nF, nF, nT))
    q = np.tile(q_in[:, None], (1, nT))
    z = np.zeros((nF, nT))

    tot_res = q_in.sum()
    # Check for feasibility
    for t in range(nT):
        if tot_res < sum(d[c, t] for c in range(nC)):
            logging.error(f"Not enough resources for demand at time {t}. Problem is infeasible.")
            exit()

    # TODO: parallelizza il loop per ogni t \in T eseguendo il codice attuale
    # oppure sempre parallelizzando il loop per ogni t \in T, ogni singolo T è un problema di flusso
    # aprendo ogni facility, che può essere risolto all'ottimo in tempo polinomiale
    # networkx
    for t in range(nT):
        for c in range(nC):
            already_assigned = 0

            f = 0
            residual_capacity = q_in[f] - sum(x[f, k, t] for k in range(nC))

            while already_assigned < d[c, t]:
                if residual_capacity > d[c, t] - already_assigned:
                    x[f, c, t] += d[c, t] - already_assigned
                    already_assigned = d[c, t]
                else:
                    if residual_capacity > 0:
                        x[f, c, t] += residual_capacity
                        already_assigned += residual_capacity
                    f += 1
                    if f < nF:
                        residual_capacity = q_in[f] - sum(x[f, k, t] for k in range(nC))

    for t in range(nT):
        for j in range(nC):
            if x[:, j, t].sum() != 0:
                for i in range(nF):
                    x[i, j, t] = x[i, j, t] / x[:, j, t].sum()

    return x, y, w, q, z

class HeuristicRunner:
    """Wrapper to run the C++ construction heuristic and handle results."""
    
    def __init__(self, executable_path):
        """
        Initialize the runner.
        
        Args:
            executable_path: Path to the compiled C++ executable
        """
        self.executable_path = Path(executable_path)
        
        if not self.executable_path.exists():
            raise FileNotFoundError(f"Executable not found: {executable_path}")
        
        if not os.access(self.executable_path, os.X_OK):
            raise PermissionError(f"Executable is not executable: {executable_path}")
    
    def create_input_file(self, nF, nC, nT, d, q_in, input_file):
        """
        Create input JSON file for the C++ program.
        
        Args:
            nF: Number of facilities
            nC: Number of customers
            nT: Number of time periods
            d: Demand matrix (nC x nT)
            q_in: Initial capacity vector (nF)
            input_file: Path where to save the input JSON
        """
        input_data = {
            "nF": nF,
            "nC": nC,
            "nT": nT,
            "d": d.tolist(),
            "q_in": q_in.tolist()
        }
        
        with open(input_file, 'w') as f:
            json.dump(input_data, f, indent=2)
        
        print(f"Input file created: {input_file}")
    
    def _run(self, input_file, output_file, num_threads=None, timeout=None) -> dict:
        """
        Run the C++ heuristic executable.
        
        Args:
            input_file: Path to input JSON file
            output_file: Path where output JSON will be written
            num_threads: Number of OpenMP threads (optional)
            timeout: Maximum execution time in seconds (optional)
            
        Returns:
            dict: Result containing status and execution info
        """
        # Build command
        cmd = [
            str(self.executable_path),
            '--input', str(input_file),
            '--output', str(output_file)
        ]
        
        if num_threads is not None:
            cmd.extend(['--threads', str(num_threads)])
        
        print(f"Running command: {' '.join(cmd)}")
        
        try:
            start_time = time.time()
            
            # Run the C++ executable
            result = subprocess.run(
                cmd,
                capture_output=True,
                text=True,
                timeout=timeout,
                check=True
            )
            
            elapsed_time = time.time() - start_time
            
            print(f"Execution completed successfully in {elapsed_time:.2f} seconds")
            print(f"stdout:\n{result.stdout}")
            
            if result.stderr:
                print(f"stderr:\n{result.stderr}")
            
            return {
                'success': True,
                'elapsed_time': elapsed_time,
                'return_code': result.returncode,
                'stdout': result.stdout,
                'stderr': result.stderr
            }
            
        except subprocess.TimeoutExpired:
            print(f"Error: Execution timed out after {timeout} seconds")
            return {
                'success': False,
                'error': 'timeout',
                'timeout': timeout
            }
            
        except subprocess.CalledProcessError as e:
            print(f"Error: Execution failed with return code {e.returncode}")
            print(f"stdout:\n{e.stdout}")
            print(f"stderr:\n{e.stderr}")
            return {
                'success': False,
                'error': 'execution_failed',
                'return_code': e.returncode,
                'stdout': e.stdout,
                'stderr': e.stderr
            }
            
        except Exception as e:
            print(f"Unexpected error: {str(e)}")
            return {
                'success': False,
                'error': 'unexpected',
                'message': str(e)
            }
    
    def _read_output(self, output_file : str) -> dict:
        """
        Read the output JSON file produced by the C++ program.
        
        Args:
            output_file: Path to the output JSON file
            
        Returns:
            dict: Dictionary containing x, y, w, q, z arrays
        """
        try:
            with open(output_file, 'r') as f:
                output_data = json.load(f)
            
            # print(f"Output file read successfully: {output_file}")
            
            # Convert to numpy arrays if numpy is available
            
            output_data = {
                'x': np.array(output_data['x']),
                'y': np.array(output_data['y']),
                'w': np.array(output_data['w']),
                'q': np.array(output_data['q']),
                'z': np.array(output_data['z'])
            }
            # print("Output converted to numpy arrays")
        
            return output_data
            
        except FileNotFoundError:
            logging.error(f"Error: Output file not found: {output_file}")
            return None
        except json.JSONDecodeError as e:
            logging.error(f"Error: Invalid JSON in output file: {e}")
            return None
        except Exception as e:
            logging.error(f"Error reading output: {str(e)}")
            return None
    
    def run_and_get_results(self, input_file, output_file, num_threads=None, timeout=None):
        """
        Convenience method: run the executable and return the results.
        
        Args:
            input_file: Path to input JSON file
            output_file: Path where output JSON will be written
            num_threads: Number of OpenMP threads (optional)
            timeout: Maximum execution time in seconds (optional)
            
        Returns:
            tuple: (execution_info, results) or (execution_info, None) if failed
        """
        execution_info = self._run(input_file, output_file, num_threads, timeout)
        
        if execution_info['success']:
            results = self._read_output(output_file)
            return execution_info, results
        else:
            return execution_info, None

def construction_heuristic_cpp(nF, nC, nT, d, q_in):
    input_file = 'input.json'
    output_file = 'output.json'
    hr = HeuristicRunner(EXECUTABLE_PATH)
    hr.create_input_file(nF, nC, nT, d, q_in, input_file)
    _, hc = hr.run_and_get_results(input_file, output_file)

    return hc["x"], hc["y"], hc["w"], hc["q"], hc["z"]

def toy_example_cpp():
    """Example usage of the HeuristicRunner."""
    
    # Configuration
    executable_path = EXECUTABLE_PATH  # Path to your compiled C++ executable
    input_file = 'input.json'
    output_file = 'output.json'
    
    # Example problem data
    nF = 3
    nC = 4
    nT = 5
    d = [
        [10.0, 12.0, 15.0, 11.0, 13.0],
        [8.0, 9.0, 11.0, 10.0, 12.0],
        [15.0, 14.0, 16.0, 13.0, 15.0],
        [7.0, 8.0, 9.0, 8.0, 10.0]
    ]
    q_in = [50.0, 45.0, 40.0]
    
    try:
        # Create runner
        runner = HeuristicRunner(executable_path)
        
        # Create input file
        runner.create_input_file(nF, nC, nT, d, q_in, input_file)
        
        # Run and get results
        exec_info, results = runner.run_and_get_results(
            input_file=input_file,
            output_file=output_file,
            num_threads=4,
            timeout=60  # 60 seconds timeout
        )
        
        if results is not None:
            print("\n=== Results ===")
            print(f"x shape: {len(results['x'])} x {len(results['x'][0])} x {len(results['x'][0][0])}")
            print(f"y shape: {len(results['y'])} x {len(results['y'][0])}")
            print(f"Execution time: {exec_info['elapsed_time']:.2f} seconds")
            
            # Example: access specific values
            print(f"\nSample values from x[0][0]: {results['x'][0][0][:3]}")
        else:
            print("\nFailed to get results")
            sys.exit(1)
    
    except Exception as e:
        print(f"Error: {str(e)}")
        sys.exit(1)

def main():    
    amelia_data = toy.dataset_amelia()

    start_t = time.time()
    x, y, w, q, z = construction_heuristic(amelia_data.nF, amelia_data.nC, amelia_data.nt, amelia_data.d, amelia_data.q_in)
    print(f"ending python implementation in {time.time() - start_t}")

    start_t = time.time()
    input_file = 'input.json'
    output_file = 'output.json'
    hr = HeuristicRunner(EXECUTABLE_PATH)
    hr.create_input_file(amelia_data.nF, amelia_data.nC, amelia_data.nt, amelia_data.d, amelia_data.q_in, input_file)
    _, hc = hr.run_and_get_results(input_file, output_file)
    print(f"ending c++ implementation in {time.time() - start_t}")


    print("check")

if __name__ == "__main__":
    main()
